Runbook: Secrets

confused-Techie opened the pull request:
As suggested on Discord we really ought to start documenting all the knowledge that currently only exists in people's head.

Seems best to do this via some simple runbooks, with my first entry being that about `secrets`.

This initial guide focusing on what secrets we use, where they are, who owns them, and what to do when updating them.

Please feel free to provide any feedback possible, as this is a first of this format.

---

As they requested this initially @savetheclocktower

confused-Techie committed:
Add runbooks dir

confused-Techie committed:
Create secrets runbook

confused-Techie committed:
Resolve formatting mistake

savetheclocktower commented on guides/runbooks/secrets.md:
```diff
@@ -0,0 +1,465 @@
+
+**IMPORTANT**
+
+The secret has very limited 'visibility', only being visible to the repositories
```
```suggestion
The secret has very limited _visibility_, only being visible to the repositories
```

savetheclocktower commented on guides/runbooks/secrets.md:
```diff
@@ -0,0 +1,465 @@
+# Secrets
```
In general: maybe soft-wrap these paragraphs so they're easier to edit?

savetheclocktower commented on guides/runbooks/secrets.md:
```diff
@@ -0,0 +1,465 @@
+
+The secret has very limited 'visibility', only being visible to the repositories
+we know need to use it, and any new repo **must** be given access to this secret
+via it's 'visibility'.
```
```suggestion
via its own “visibility” setting.
```

savetheclocktower commented on guides/runbooks/secrets.md:
```diff
@@ -0,0 +1,465 @@
+
+**DESCRIPTION**
+
+The AppleID of a developer Apple account, that allows us to sign macOS binaries produced
```
```suggestion
The AppleID of a developer Apple account. Allows us to sign macOS binaries produced
```

savetheclocktower commented on guides/runbooks/secrets.md:
```diff
@@ -0,0 +1,465 @@
+
+### 1.1 `NPM_PUBLISH_TOKEN`
+
+**DESCRIPTION**
```
```suggestion
#### Description
```

In general these could all be headings instead of random bold words.

savetheclocktower commented on guides/runbooks/secrets.md:
```diff
@@ -0,0 +1,465 @@
+
+A GitHub PAT token with zero permissions. This token is used so that CirrusCI's
+API requests to GitHub are authenticated and have a higher API rate limit.
+This was done due to a high number of requests when downloading 'RipGrep'.
```
```suggestion
This was done due to a high number of requests when downloading `ripgrep`.
```

savetheclocktower reviewed (approved):
All in all, this looks great. I started to leave some suggestions (feel free to take or leave any of them) but just having this stuff written down is great and I'm not going to stress out about formatting or structure. Thanks!

DeeDeeG commented on guides/runbooks/secrets.md:
```diff
@@ -0,0 +1,465 @@
+**OWNERSHIP**
+
+- Last Updated: August 2025
+- Who: Unknown
```
```suggestion
- Who: DeeDeeG
```

DeeDeeG commented on guides/runbooks/secrets.md:
```diff
@@ -0,0 +1,465 @@
+**OWNERSHIP**
+
+- Last Updated: August 2025
+- Who: Unknown
```
```suggestion
- Who: DeeDeeG
```

DeeDeeG commented on guides/runbooks/secrets.md:
```diff
@@ -0,0 +1,465 @@
+**OWNERSHIP**
+
+- Last Updated: August 2025
+- Who: Unknown
```
```suggestion
- Who: DeeDeeG
```

DeeDeeG commented on guides/runbooks/secrets.md:
```diff
@@ -0,0 +1,465 @@
+
+PAT tokens by default have an expiration date, and if/when this token expires it
+**must** be updated by creating a new CirrusCI 'Secured Variable' and modifying
+the variable within `.cirrus.yml` in `pulsar-edit/pulsar`.
```
I've managed to find an easier way to update this PAT across the entire core repo for Cirrus purposes, without needing to do a PR for it on multiple branches (e.g. `master` and `updated-electron-next`).

(I'm not confident to do it with permission-having tokens like the Rolling or Electron Next upload ones, but for the permissionless `GITHUB_TOKEN` I've been comfortable with doing it this way. It's an option.) 

See this closed PR I made to update documentation of the variable in core repo https://github.com/pulsar-edit/pulsar/pull/1235. (I closed it because I felt it had stalled out, was hard to explain to the rest of the team, and because I was shy to poke anyone further about it to get a yea/nay.)

```suggestion
**must** be updated by creating a new CirrusCI 'Secured Variable' and either
modifying the variable within `.cirrus.yml` in `pulsar-edit/pulsar`, or updating
the 'Environment variable overrides' entry for `GITHUB_TOKEN` in the Cirrus CI
web UI.
```

confused-Techie replied:
This is a super smart idea honestly. And I'd fully support you opening back up that PR and know you're more than free to poke me if needed about it. But regardless having this comment here is a good call

DeeDeeG reviewed (commented):
I reckon the "Last Updated" info, while helpful now, may soon be outdated. I don't expect to impose on those who update PATs or credentials regularly that they also swing around and update this document, so I propose to delete the "Last updated" lines.

(I reckon those of us preparing or reviewing this PR know the dates written there now, and that by the time they would be out of date, they might only serve to confuse or misdirect others not present at this PR.)

Also proposed some diffs to add what I know about the tokens I've been presiding over for Cirrus.

confused-Techie commented:
@savetheclocktower Thanks a ton for your feedback, I'll address some of the changes suggested of data before getting to those formatting changes, but they do look like good calls.

And to @DeeDeeG the `Last Updated` datapoint is less meant to indicate when the secret itself was last updated, but when the ownership of said secret was last updated in this document. With the idea that if someone reviews this document a year from now and realize they now have ownership of something or that someone else days they could update the ownership datapoint along with when they are updating it. 
But seeing as how you found it confusing I'll be sure to change it. 
Do you think this is still valuable data? By maybe changing the wording to say "Accurate as of" instead of "Last Updated"? Or do you think it's best to remove it all together like you suggested?

confused-Techie committed:
Update guides/runbooks/secrets.md

Co-authored-by: DeeDeeG <DeeDeeG@users.noreply.github.com>

confused-Techie committed:
Update guides/runbooks/secrets.md

Co-authored-by: Andrew Dupont <andrew@andrewdupont.net>

confused-Techie committed:
Update guides/runbooks/secrets.md

Co-authored-by: Andrew Dupont <andrew@andrewdupont.net>

confused-Techie committed:
Update guides/runbooks/secrets.md

Co-authored-by: Andrew Dupont <andrew@andrewdupont.net>

confused-Techie committed:
Update guides/runbooks/secrets.md

Co-authored-by: Andrew Dupont <andrew@andrewdupont.net>

confused-Techie committed:
Update guides/runbooks/secrets.md

Co-authored-by: DeeDeeG <DeeDeeG@users.noreply.github.com>

confused-Techie committed:
Update guides/runbooks/secrets.md

Co-authored-by: DeeDeeG <DeeDeeG@users.noreply.github.com>

confused-Techie committed:
Update guides/runbooks/secrets.md

Co-authored-by: DeeDeeG <DeeDeeG@users.noreply.github.com>

confused-Techie committed:
Update guides/runbooks/secrets.md

Co-authored-by: Andrew Dupont <andrew@andrewdupont.net>

confused-Techie committed:
Formatting: Replace `**WORD**` with headers

confused-Techie committed:
Remove hard-wrap on all text blocks

confused-Techie commented:
I've now addressed all feedback provided if anyone would like to re-review so we can get this merged, I'd appreciate it!
@DeeDeeG @savetheclocktower

confused-Techie commented:
Actually, @savetheclocktower has already provided an approval. Considering that everyone's concerns have been handled, and the propensity of PR's in this repository to hang in limbo, I'm gonna go ahead and run with that original approval and merge this one.

Thanks for everyone that took a look at this, appreciate it!

confused-Techie committed:
Change `Last Updated` to `Accurate as of`

This is meant to address feedback and attempt to better show what this datapoint is supposed to represent. We can always consider fully removing this in the future if we find it pointless to keep.

savetheclocktower commented on guides/runbooks/secrets.md:
```diff
@@ -0,0 +1,411 @@
+
+#### Description
+
+This token allows publish access to our `@pulsar-edit` NPM account. Likely this token was generated by a user on the NPM organization that has publish access.
```
I should've asked this before this landed, but: how is membership in the NPM organization decided? Do we have a bus factor there as well?

confused-Techie replied:
Not as much of one, users have to be added to the NPM organization.

But that's actually perfect subject matter for #319 and I can go ahead and add it there. Thanks for pointing it out
